### PR TITLE
Adding $flatten

### PR DESCRIPTION
**Pull request by TheNavigat**

### CODE REVIEW

File: specification.yml [hunk 1]
```diff
@@ -258,6 +258,24 @@ error: true
...
+---
+title:    flatten empty array
+context:  {}
+template: {$flatten: []}
+error: true
+---
+title:    flatten empty array of arrays
+context:  {}
+template: {$flatten: [[]]}
+result:   []
```
Comment (djmitche) [inline specification.yml]: I like this!  A few error conditions would be good, too: flattening arrays of various other types, or not giving an array to flatten at all.

File: specification.yml [hunk 2]
```diff
@@ -268,12 +268,22 @@ result:   [1, 2, 3, 4, 5]
...
+error:    true
 ---
 title:    flatten empty array of arrays
 context:  {}
 template: {$flatten: [[]]}
 result:   []
+---
+title:    flatten nothing
+context:  {}
+template: {$flatten: }
```
Comment (djmitche) [inline specification.yml]: Is that even legitimate YAML?  We don't want to test the YAML parser :)

What I meant was something like `{$flatten: 13}` or `{$flatten: [13]}` or `{$flatten: {a: 10, b: 20}}`.

File: specification.yml [hunk 3]
```diff
@@ -268,12 +268,22 @@ result:   [1, 2, 3, 4, 5]
...
+error:    true
+---
+title:    flatten mixed types
+context:  {}
+template: {$flatten: [[1, 2], [['a', 'b']], [{}], [[]]]}
+result:   [1, 2, ['a', 'b'], {}, []]
```
Comment (djmitche) [inline specification.yml]: I like this :)

File: specification.yml (hunk 2 above)
Comment (djmitche) [inline specification.yml]: Haha, it is legitimate YAML (equivalent to `{$flatten: null}`)

Comment (TheNavigat) [inline specification.yml]: The reason for why it returns an error is different, though. It's because we can't parse null.

Comment (djmitche) [inline specification.yml]: Right, and that's a perfectly valid thing to test.  I did change it to say "null" explicitly in my follow-up.

**Pull request merged**